src/hdmf/common/hierarchicaltable.py

```diff
@@ -0,0 +1,258 @@
+    DataFrame include:
+
+    * pandas.DataFrame.reset_index to turn the data from the pandas.MultiIndex into columns
+    * :py:meth:`~hdmf.common.hierarchicaltable.drop_id_columns` to remove all 'id' columns
+    * :py:meth:`~hdmf.common.hierarchicaltable.flatten_column_index` to flatten the column index
+    """
+    # TODO: Need to deal with the case where we have more than one DynamicTableRegion column in a given table
+    # Get the references column
+    foreign_columns = dynamic_table.get_foreign_columns()
+    if len(foreign_columns) == 0:
+        return dynamic_table.to_dataframe()
+    hcol_name = foreign_columns[0]  # We only denormalize the first foreign column for now
+    hcol = dynamic_table[hcol_name]
+    # If our foreign column is a vectorindex
+    hcol_target = hcol.table if isinstance(hcol, DynamicTableRegion) else hcol.target.table
+
+    # Create the data variables we need to collect the data for our output dataframe and associated index
+    index = []
+    data = []
+    columns = None
+    index_names = None
+
+    # If we have indexed columns (other than our hierarchical column) then our index data for our
+    # MultiIndex will contain lists as elements (which are not hashable) and as such create an error.
+    # As such we need to check if we have any affected columns so we can  fix our data
+    # indexed_column_indicies = np.where([isinstance(dynamic_table[colname], VectorIndex)
+    #                                    for colname in dynamic_table.colnames if colname != hcol_name])[0]
+    # indexed_column_indicies += 1  # Need to increment by 1 since we add the row id in our iteration below
+
+    # Case 1:  Our DynamicTableRegion column points to a regular DynamicTable
+    #          If this is the case than we need to de-normalize the data and flatten the hierarchy
+    if not hcol_target.has_foreign_columns():
+        # 1) Iterate over all rows in our hierarchical columns (i.e,. the DynamicTableRegion column)
+        #    First we here get a list of DataFrames, one for each row. If our hcol is a VectorIndex
+        #    (i.e., out column is a ragged array), then simply loading the data from the VectorIndex
+        #    will do the trick. If we have a regular VectorData column, then we need to load the
+        #    elements ourselves (using slice syntax to make sure we get DataFrames) one-row-at-a-time
+        if isinstance(hcol, VectorIndex):
+            rows = hcol.get(slice(None), index=False, df=True)
+        else:
+            rows = [hcol[i:(i+1)] for i in range(len(hcol))]
+        # Now we iterate over the rows, where each row is described by a DataFrame with one-or-more rows
+        for row_index, row_df in enumerate(rows):
+            # 1.1): Since hcol is a DynamicTableRegion, each row returns another pandas.DataFrame so we
+            #       next need to iterate over all rows in that table to denormalize our data
```

Comment:
There is no 1.2 and no guarantee that hcol is a DynamicTableRegion. Can this comment be removed and the lower numbers updated from "1.1.1" to "1.1"?
```suggestion
            # Iterate over all rows in that table to denormalize our data
```

Reply:
Ok, I'll update the comments. The original logic assumed that hcol was a VectorIndex (which is true for the ICEPhys tables) but depending on the structure of the tables hcol can either be a DynamicTableRegion or a VectorIndex that points to a DynamicTableRegion. This is being dealt with before the loop now when constructing the rows array.

Reply:
I updated the comments throughout the to_hierarchical_dataframe function accordingly.